### ISSUE TITLE
showModal throws exception when dialog is inside a custom-element

### ISSUE THREAD

**Issue by akc42** [GoogleChrome/dialog-polyfill#186]
I have a large web-components (using lit-element) application.  It will be used in a very closed environment, so I am able to control the browser support.  So far its been used exclusively with chrome.

I do use the `<dialog>` element inside a few components, and built in the use of the polyfill from the beginning.  These elements exclusively use `showModal()` to open dialogs when needed.

I have been asked to check for IOS support on ipads - but with them it has been impossible to debug why things didn't work, so I just starting using Safari on the mac.

However using the debugger on my mac with Safari I have discovered that `showModal()` tries to check whether the dialog exists by doing a `document.body.contains(this.dialog_)` call.  `Node.contains()` doesn't include elements inside the shadowRoot of a contained element, so the call throws an exception.

**Comment by samthor** [comment]
Thanks for the report. The polyfill should work inside a shadow root (even if automatically injecting the CSS is up in the air). I'll look at this soon.

**Comment by paulerickson** [comment]
Removing those checks, it seems to [work for me](https://github.com/paulerickson/dialog-polyfill/commit/fdf449462052662048805090d75561239bd7490e#diff-168726dbe96b3ce427e7fedce31bb0bc) ([CDN dist here](https://cdn.jsdelivr.net/gh/paulerickson/dialog-polyfill@master/dist/dialog-polyfill.esm.js)), except on form submission the page reloads, but maybe that's unrelated.

**Comment by nalply** [comment]
An alternative to remove the check is using something like this (untested) at:

https://github.com/GoogleChrome/dialog-polyfill/blob/master/index.js#L287

```javascript
if (!document.body.contains(this.dialog_)) {
    const element = this.dialog_;
    while (element != null && element.nodeType != Node.DOCUMENT_FRAGMENT_NODE) {
        element = element.parentNode;
    }
    if (!element.contains(this.dialog_))
        throw new Error('Failed to execute \'showModal\' on dialog: The element is not in a Document.');
}
```

This walks up the DOM Tree till it finds the shadow root host (`DOCUMENT_FRAGMENT_NODE`), then retries the same check from the host instead of body.

**Comment by hadriann** [comment]
@samthor @nalply can the nalply/dialog-polyfill fork be merged to make the dialog usable inside shadow DOM?

**Comment by samthor** [comment]
I'll look at this soon.

My concern is that when you're in a shadow root, it's very likely that
something is causing layout, so the `position: fixed` code won't work. But
we detect that anyway and complain if you're doing something wrong :)

**Comment by hadriann** [comment]
@samthor @nalply I think that any layout issue inside a shadow root is acceptable compared to not working at all in that context. Would be nice and helpful to see that fix merged. Thanks :)

**Comment by kwaclaw** [comment]
Whats the best way forward? Looks like this is not on anyone's list to fix.
Should we use the  nalply/dialog-polyfill fork? Or is another polyfill out there?

**Comment by samthor** [comment]
If there's a PR that fixes this I can take a look very easily.

**Comment by hadriann** [comment]
I just forked and will try to do the minimal changes for this issue to be fixed. @samthor so you know why the `dist` files in the repo seem not to be up-to-date with `index.js`?

**Comment by samthor** [comment]
I'm not sure, but I can build it with Closure if needed. IIRC the non-dist
version should be fine for testing.

**Comment by hadriann** [comment]
@samthor please check pull request https://github.com/GoogleChrome/dialog-polyfill/pull/194 that fixes this issue. Thanks!

**Comment by samthor** [comment]
I've published 0.5.1 which includes this.

**Comment by hadriann** [comment]
@samthor I believe this issue can be closed as per 0.5.1 ?

**Comment by samthor** [comment]
I bumped up `package.json` to 0.5.1.
